src/connector/src/sink/formatter/mod.rs

```diff
@@ -0,0 +1,52 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use itertools::Itertools;
+use risingwave_common::array::{Op, StreamChunk};
+use risingwave_common::row::Row;
+
+use crate::sink::Result;
+
+mod append_only;
+mod upsert;
+
+pub use append_only::AppendOnlyFormatter;
+pub use upsert::UpsertFormatter;
+
+/// Transforms a `StreamChunk` into a sequence of key-value pairs according a specific format,
+/// for example append-only, upsert or debezium.
+///
+/// Each row in the `StreamChunk` may emit 0, 1, or more output pairs, which would be concatenated
+/// (flattened) by [`format_chunk`]. This auxiliary function is not included as a provided method
+/// due to some typing issue.
+///
+/// Generator syntax, once available, may make this simpler. Note that we only need an iterator,
+/// and async generator would be an overkill.
+pub trait SinkFormatter {
+    type K;
+    type V;
+    type I: IntoIterator<Item = (Option<Self::K>, Option<Self::V>)>;
+
+    fn format_row(&mut self, op: Op, row: impl Row + Copy) -> Result<Self::I>;
```

Comment:
Totally agree with you about the ideal case: using generator to do a `format_chunk` directly. This has been acknowledged in the doc comments of `SinkFormatter`. However, I see no existing usage of `from_generator` in our codebase and I am not sure if it is okay to adopt it. We do have `futures_async_stream::try_stream` but its `async` is unnecessary.

> The current `format_row` assumes that the each output message is generated within the scope of a single row.

Not really. This is why `format_row` takes `&mut self` rather than `&self`. States that need to carry over rows can be stored in `&mut self`. You can refer to the preview PR #12125 to see how debezium works in this case. (It was not included in this PR due to its length.) This is basically managing closure states of a generator between `yield` manually.

> can we assume that key always exists

I had the same question when I work on this code. But I consider it to be orthogonal to this refactor. For nats, I plan to have a `EmptyEncoder<Output=()>` as key encoder. In other cases an empty string or bytes may also be enough. cc @tabVersion

Reply:
> However, I see no existing usage of from_generator in our codebase and I am not sure if it is okay to adopt it. We do have futures_async_stream::try_stream but its async is unnecessary.

I think it's fine to use it. Actually the code generated from the try_stream macro is also a generator. I just had a try on modifying `gen_debezium_message_stream` to return a iterator created from `from_generator`. It's doable.

> Not really. This is why format_row takes &mut self rather than &self. States that need to carry over rows can be stored in &mut self

Agree that we can manage the state with `&mut self`. But if we can use generator, we can avoid this and make our code more readable and intuitive.

Reply:
Latest on `Self::K` vs `Option<Self::K>`: For append-only without key, we want kafka to partition with round-robin. The current assumption of guaranteed `Some("{}")` as key is wrong and leads to data skew #9443